WIP: HTML Reporter: Fuzzy search using fuse@6.1.0-alpha.1

ventuno opened the pull request:
Follow up on: https://github.com/qunitjs/qunit/pull/1442#issuecomment-643550086. Testing `Fuse.js`@6.1.0-alpha.1.

#### `Fuse.js`@6.1.0-alpha.1.

Search: "assert".
<img width="408" alt="fuse@6 1 0-alpha 1  - assert" src="https://user-images.githubusercontent.com/5890858/84600607-6508e880-ae2f-11ea-8ac4-ec271826fb1d.png">

Search: "beforeEach".
<img width="407" alt="fuse@6 1 0-alpha 1  - beforeEach" src="https://user-images.githubusercontent.com/5890858/84600606-5e7a7100-ae2f-11ea-9c16-b2be66044b06.png">

Search: "promiseaware".
<img width="409" alt="fuse@6 1 0-alpha 1  - promiseaware" src="https://user-images.githubusercontent.com/5890858/84600601-57ebf980-ae2f-11ea-844d-ef49b41d6a91.png">

Search: "promiseawarebeforeEach".
<img width="409" alt="fuse@6 1 0-alpha 1  - promiseawarebeforeEach" src="https://user-images.githubusercontent.com/5890858/84600596-4b67a100-ae2f-11ea-9106-cc24073bb04c.png">

Search: "script".
<img width="406" alt="fuse@6 1 0-alpha 1  - script" src="https://user-images.githubusercontent.com/5890858/84600589-4145a280-ae2f-11ea-90da-af2f7a9d42b2.png">


#### `Fuse.js`@6.1.0-alpha.0.

Search: "assert".
<img width="410" alt="fuse@6 1 0-alpha 0  - assert" src="https://user-images.githubusercontent.com/5890858/84580187-8f07cf80-ad89-11ea-8d7a-be1dfc97ef87.png">

Search: "beforeEach".
<img width="408" alt="fuse@6 1 0-alpha 0  - beforeEach" src="https://user-images.githubusercontent.com/5890858/84580189-90d19300-ad89-11ea-9661-40c197ef66f9.png">

Search: "promiseaware".
<img width="407" alt="fuse@6 1 0-alpha 0  - promiseaware" src="https://user-images.githubusercontent.com/5890858/84580191-9202c000-ad89-11ea-80d8-cdcdab65aec6.png">

Search: "promiseawarebeforeEach".
<img width="409" alt="fuse@6 1 0-alpha 0  - promiseawarebeforeEach" src="https://user-images.githubusercontent.com/5890858/84580192-929b5680-ad89-11ea-993c-90ca6d87741c.png">

Search: "script".
<img width="408" alt="fuse@6 1 0-alpha 0  - script" src="https://user-images.githubusercontent.com/5890858/84580194-929b5680-ad89-11ea-968a-b4c2e1e1b257.png">

krisk commented:
Re: https://github.com/qunitjs/qunit/pull/1442#issuecomment-643686228

The reason is because of the [field-length norm](https://fusejs.io/concepts/scoring-theory.html#field-length-norm), where the shorter the field is, the higher the relevance of a match will be.

However, it sounds like what you may be looking for then is a basic relevance scoring. In this case, I'd recommend pulling `fuse.js@6.1.0-alpha.1`, and setting these options:

- `ignoreFieldNorm` => `true`
- `ignoreLocation` => `true`

By doing this, when I search for `"promiseawarebeforeEach"`, I get the following:

```js
[
  ({
    item: 'Module with Promise-aware beforeEach',
    refIndex: 3,
    score: 0.09090909090909091
  },
  {
    item: 'Module with Promise-aware afterEach',
    refIndex: 5,
    score: 0.2727272727272727
  },
  {
    item: 'Promise-aware return values without beforeEach/afterEach',
    refIndex: 4,
    score: 0.4090909090909091
  },
  { item: 'async beforeEach test', refIndex: 1, score: 0.5 },
  { item: 'assert.async in beforeEach', refIndex: 2, score: 0.5 },
  { item: 'beforeEach', refIndex: 0, score: 0.5454545454545454 })
]
```

Hopefully, this settles the use case? 😅

ventuno commented:
Thanks @krisk for further looking into this. I updated to `fuse.js@6.1.0-alpha.1` and although results look better I think they still look a little odd (see updated screenshots in the description. I kept both sets for comparison).

[Here's the list of modules we are filtering](https://pastebin.com/RGzFjqgq) I removed all other properties and only left `name`, which is the property we use for filtering. It's a little hard for me to provide the expected sorting, but my expectation would be that if I search "promiseawarebeforeEach" all results containing "Promise-aware" and "beforeEach" (together) should be closer to the top, so for example seeing "Module with Promise-aware afterEach" ranked higher than "Promise-aware return values without beforeEach/afterEach" is not really what I expect (`fuzzysort` is very close to what I expect [example 1](https://github.com/qunitjs/qunit/pull/1440#issue-418914860) and [example 2](https://github.com/qunitjs/qunit/pull/1442#issuecomment-636652146)).

As a I user, I know what my module name is and I'm just trying to get to it using the filter.
As far as I understand (just based on empirical observation) `Fuse.js` is optimized to surface matches when I _don't know_ what's in the content, while [`fuzzysort`](https://github.com/farzher/fuzzysort)'s goal seems more optimal to our goal as it strives to match SublimeText's file finder.
Does that make sense?

ventuno commented:
Closing based on the above.
